File: javascript/packages/core/src/lint-result.ts
import { Location, type SerializedLocation } from "./location.js"

import type { DiagnosticSeverity, DiagnosticTag } from "./diagnostic.js"

export interface SerializedLintOffense {
  rule: string
  code: string
  source: string
  message: string
  severity: DiagnosticSeverity
  location: SerializedLocation
  tags?: DiagnosticTag[]
}

export interface SerializedAutofixResult {
  source: string
  fixed: SerializedLintOffense[]
  unfixed: SerializedLintOffense[]
}

export interface SerializedLintResult {
  offenses: SerializedLintOffense[]
  errors: number
  warnings: number
  info: number
  hints: number
  ignored: number
}

export class BackendLintOffense {
  readonly rule: string
  readonly code: string
  readonly source: string
  readonly message: string
  readonly severity: DiagnosticSeverity
  readonly location: Location
  readonly tags?: DiagnosticTag[]

  constructor(
    rule: string,
    code: string,
    source: string,
    message: string,
    severity: DiagnosticSeverity,
    location: Location,
    tags?: DiagnosticTag[],
  ) {
    this.rule = rule
    this.code = code
    this.source = source
    this.message = message
    this.severity = severity
    this.location = location
    this.tags = tags
  }

  static from(offense: SerializedLintOffense): BackendLintOffense {
    return new BackendLintOffense(
      offense.rule,
      offense.code,
      offense.source,
      offense.message,
      offense.severity,
      Location.from(offense.location),
      offense.tags,
    )
  }

  get isError(): boolean {
    return this.severity === "error"
  }

  get isWarning(): boolean {
    return this.severity === "warning"
  }

  get isInfo(): boolean {
    return this.severity === "info"
  }

  get isHint(): boolean {
    return this.severity === "hint"
  }

  toHash(): SerializedLintOffense {
    return {
      rule: this.rule,
      code: this.code,
      source: this.source,
      message: this.message,
      severity: this.severity,
      location: this.location.toHash(),
      ...(this.tags ? { tags: this.tags } : {}),
    }
  }

  toJSON(): SerializedLintOffense {
    return this.toHash()
  }

  inspect(): string {
    return `#<Herb::BackendLintOffense rule=${JSON.stringify(this.rule)} severity=${JSON.stringify(this.severity)} message=${JSON.stringify(this.message)} location=${this.location.treeInspect()}>`
  }

  toString(): string {
    return this.inspect()
  }
}

export class BackendLintResult {
  readonly offenses: BackendLintOffense[]
  readonly errors: number
  readonly warnings: number
  readonly info: number
  readonly hints: number
  readonly ignored: number

  constructor(
    offenses: BackendLintOffense[],
    errors: number,
    warnings: number,
    info: number,
    hints: number,
    ignored: number,
  ) {
    this.offenses = offenses
    this.errors = errors
    this.warnings = warnings
    this.info = info
    this.hints = hints
    this.ignored = ignored
  }

  static from(result: SerializedLintResult): BackendLintResult {
    const offenses = (result.offenses || []).map((offense) => BackendLintOffense.from(offense))

    return new BackendLintResult(
      offenses,
      result.errors || 0,
      result.warnings || 0,
      result.info || 0,
      result.hints || 0,
      result.ignored || 0,
    )
  }

  get offenseCount(): number {
    return this.offenses.length
  }

  get clean(): boolean {
    return this.offenses.length === 0
  }

  toHash(): SerializedLintResult {
    return {
      offenses: this.offenses.map((offense) => offense.toHash()),
      errors: this.errors,
      warnings: this.warnings,
      info: this.info,
      hints: this.hints,
      ignored: this.ignored,
    }
  }

  toJSON(): SerializedLintResult {
    return this.toHash()
  }

  inspect(): string {
    return `#<Herb::BackendLintResult offenses=${this.offenseCount} errors=${this.errors} warnings=${this.warnings} info=${this.info} hints=${this.hints}>`
  }

  toString(): string {
    return this.inspect()
  }
}

export class BackendAutofixResult {
  readonly source: string
  readonly fixed: BackendLintOffense[]
  readonly unfixed: BackendLintOffense[]

  constructor(source: string, fixed: BackendLintOffense[], unfixed: BackendLintOffense[]) {
    this.source = source
    this.fixed = fixed
    this.unfixed = unfixed
  }

  static from(result: SerializedAutofixResult): BackendAutofixResult {
    return new BackendAutofixResult(
      result.source,
      (result.fixed ?? []).map(BackendLintOffense.from),
      (result.unfixed ?? []).map(BackendLintOffense.from),
    )
  }
}

File: herb-linter/lib/herb/linter/lint_result.rb
# frozen_string_literal: true
# typed: true

module Herb
  class Linter
    class LintResult
      attr_reader :offenses #: Array[Offense]
      attr_reader :errors #: Integer
      attr_reader :warnings #: Integer
      attr_reader :info #: Integer
      attr_reader :hints #: Integer
      attr_reader :ignored #: Integer

      #: (Array[Offense], Integer, Integer, Integer, Integer, Integer) -> void
      def initialize(offenses, errors, warnings, info, hints, ignored)
        @offenses = offenses
        @errors = errors
        @warnings = warnings
        @info = info
        @hints = hints
        @ignored = ignored
      end

      #: (Hash[String, untyped]) -> LintResult
      def self.from(hash)
        offenses = (hash["offenses"] || []).map { |offense| Offense.from(offense) }

        new(
          offenses,
          hash["errors"] || 0,
          hash["warnings"] || 0,
          hash["info"] || 0,
          hash["hints"] || 0,
          hash["ignored"] || 0
        )
      end

      #: () -> Integer
      def offense_count
        offenses.length
      end

      #: () -> bool
      def clean?
        offenses.empty?
      end

      #: (LintResult) -> LintResult
      def merge(other)
        LintResult.new(
          offenses + other.offenses,
          errors + other.errors,
          warnings + other.warnings,
          info + other.info,
          hints + other.hints,
          ignored + other.ignored
        )
      end

      #: () -> Hash[Symbol, untyped]
      def to_hash
        {
          offenses: offenses.map(&:to_hash),
          errors: errors,
          warnings: warnings,
          info: info,
          hints: hints,
          ignored: ignored,
        }
      end

      #: (?untyped) -> String
      def to_json(state = nil)
        to_hash.to_json(state)
      end

      #: () -> String
      def inspect
        %(#<Herb::Linter::LintResult offenses=#{offense_count} errors=#{errors} warnings=#{warnings} info=#{info} hints=#{hints}>)
      end
    end
  end
end
